src/Native/Unix/System.Security.Cryptography.Native.Apple/pal_x509.c

```diff
@@ -719,3 +729,176 @@ int32_t AppleCryptoNative_X509CopyWithPrivateKey(SecCertificateRef cert,
     *pOSStatus = status;
     return status == noErr;
 }
+
+int32_t AppleCryptoNative_X509MoveToKeychain(SecCertificateRef cert,
+                                             SecKeychainRef targetKeychain,
+                                             SecKeyRef privateKey,
+                                             SecIdentityRef* pIdentityOut,
+                                             int32_t* pOSStatus)
+{
+    if (pIdentityOut != NULL)
+        *pIdentityOut = NULL;
+    if (pOSStatus != NULL)
+        *pOSStatus = noErr;
+
+    if (cert == NULL || targetKeychain == NULL || pIdentityOut == NULL || pOSStatus == NULL)
+    {
+        return -1;
+    }
+
+    SecKeychainRef curKeychain = NULL;
+    SecKeyRef importedKey = NULL;
+    OSStatus status = SecKeychainItemCopyKeychain((SecKeychainItemRef)cert, &curKeychain);
+
+    if (status == errSecNoSuchKeychain)
+    {
+        status = noErr;
+    }
+    else
+    {
+        if (curKeychain != NULL)
+        {
+            CFRelease(curKeychain);
+        }
+
+        if (status == noErr)
+        {
+            // Usage error: The certificate should have been freshly imported by the PFX loader,
+            // and therefore have no keychain.
+            return -2;
+        }
+    }
+
+    if (status == noErr && privateKey != NULL)
+    {
+        status = SecKeychainItemCopyKeychain((SecKeychainItemRef)privateKey, &curKeychain);
+
+        if (status == errSecNoSuchKeychain)
+        {
+            status = AddKeyToKeychain(privateKey, targetKeychain, &importedKey);
+        }
+        else
+        {
+            if (curKeychain != NULL)
+            {
+                CFRelease(curKeychain);
+            }
+
+            if (status == noErr)
+            {
+                // This is a usage error, the only expected call is from the PFX loader,
+                // which has an ephemeral key reference, therefore no keychain.
+                return -3;
+            }
+        }
+    }
+
+    if (status == noErr)
+    {
+        status = SecCertificateAddToKeychain(cert, targetKeychain);
+    }
+
+    if (status == noErr && privateKey != NULL)
+    {
+        CFMutableDictionaryRef query = NULL;
+        CFArrayRef searchList = NULL;
+        CFArrayRef itemMatch = NULL;
+        CFTypeRef result = NULL;
+
+        if (status == noErr)
+        {
+            query = CFDictionaryCreateMutable(
+                kCFAllocatorDefault, 0, &kCFTypeDictionaryKeyCallBacks, &kCFTypeDictionaryValueCallBacks);
+
+            if (query == NULL)
+            {
+                status = errSecAllocate;
+            }
+        }
+
+        if (status == noErr)
+        {
+            const void* constTargetKeychain = targetKeychain;
+            searchList = CFArrayCreate(NULL, (const void**)(&constTargetKeychain), 1, &kCFTypeArrayCallBacks);
```

Comment:
What is kCFTypeArrayCallBacks?  I take it that's not something that needs to be freed on failure?

Reply:
It's their standard allocator/deallocator for array elements.  The k prefix means "k"onstant, so there's no freeing that should be done.